pkg/services/ngalert/schedule/schedule.go

```diff
@@ -653,7 +637,43 @@ func (r *alertRuleRegistry) keyMap() map[models.AlertRuleKey]struct{} {
 type alertRuleInfo struct {
 	evalCh  chan *evalContext
 	stopCh  chan struct{}
-	version int64
+	mtx     sync.Mutex
+	stopped *atomic.Bool // indicates that the stopCh was closed and therefore the rule evaluation routine should exit
+}
+
+func newAlertRuleInfo() *alertRuleInfo {
+	return &alertRuleInfo{evalCh: make(chan *evalContext), stopCh: make(chan struct{}), stopped: atomic.NewBool(false)}
+}
+
+// Stop signals the rule evaluation routine to stop via stopCh and sets flag stopped to true. Does nothing If the loop is stopped
+func (a *alertRuleInfo) Stop() bool {
```

Comment:
In the past I've just done something like:

```suggestion
func (a *alertRuleInfo) Stop() bool {
     a.mtx.Lock()
     defer a.mtx.Unlock()
     if !a.stopped {
          close(a.stopCh)
          a.stopped = true
     }
     return a.stopped
}
```

Reply:
There is a pattern called "double-checked locking", which implies that the expression is evaluated before and after acquiring a lock to avoid\reduce congestion in the code wrapped by the mutex. I think in this case and this particular design of the rule evaluation routine, this approach is necessary. 
The code you suggest (I assume the same pattern you suggest using for Eval (and Update, not in this PR) as well) has a problem:
when one goroutine calls Stop, it can end up waiting for ruleRoutine to fetch data from the `stopCh`. The evaluation loop does not guarantee how fast it will read that channel because when it handles evaluation it does some pretty time-consuming operations (calls external services, retries). Therefore, the mutex can be locked for a substantial time.  
When during that time some other goroutine will try to interact with the control structure alertRuleInfo (Stop - can come from an API (other PR) as well as scheduler's sync), Update - can come from multiple concurrent requests, Eval - can come from scheduler's sync) it will be blocked on the mutex just to check whether it is closed or not. but that information is already known (it is closed at that time).

Reply:
The reason I used atomic boolean was that there is a guess that using a regular bool can make race detector complain about it.

Reply:
Oh, I just realized that I (as well as in your suggestion) do not send data to stopCh (before I sent a message as we do currently). So, yes, your suggestion does make sense.